File: go-backend-interview-prep/03_functions_and_methods/05_struct_embedding/main.go
package main

import "fmt"

// --- Base types ---
type Base struct {
	ID int
}

func (b Base) Describe() string {
	return fmt.Sprintf("Base{ID: %d}", b.ID)
}

type Logger struct {
	Prefix string
}

func (l Logger) Log(msg string) {
	fmt.Printf("[%s] %s\n", l.Prefix, msg)
}

// --- Embedding: User embeds Base ---
type User struct {
	Base        // embedded (promoted fields and methods)
	Name string
}

// --- Embedding multiple structs ---
type Admin struct {
	User          // embeds User (which embeds Base)
	Logger        // embeds Logger
	Level  string
}

// --- Field shadowing ---
type Employee struct {
	Base
	ID   string // shadows Base.ID (different type!)
	Name string
}

func main() {
	// --- Basic embedding ---
	fmt.Println("--- Basic embedding ---")
	u := User{
		Base: Base{ID: 1},
		Name: "Alice",
	}
	fmt.Println("u.ID:", u.ID)             // promoted field
	fmt.Println("u.Describe():", u.Describe()) // promoted method
	fmt.Println("u.Name:", u.Name)

	// --- Deep embedding ---
	fmt.Println("\n--- Deep embedding ---")
	a := Admin{
		User:   User{Base: Base{ID: 42}, Name: "Bob"},
		Logger: Logger{Prefix: "ADMIN"},
		Level:  "super",
	}
	fmt.Println("a.ID:", a.ID)             // promoted through User -> Base
	fmt.Println("a.Name:", a.Name)         // promoted from User
	fmt.Println("a.Describe():", a.Describe()) // promoted from Base
	a.Log("system started")               // promoted from Logger

	// --- Field shadowing ---
	fmt.Println("\n--- Field shadowing ---")
	e := Employee{
		Base: Base{ID: 100},
		ID:   "EMP-001", // shadows Base.ID
		Name: "Charlie",
	}
	fmt.Println("e.ID:", e.ID)           // "EMP-001" (outer field wins)
	fmt.Println("e.Base.ID:", e.Base.ID) // 100 (access shadowed via type name)

	// --- Embedding is NOT inheritance ---
	fmt.Println("\n--- Receiver is inner type ---")
	// u.Describe() calls Base.Describe(), not User.Describe()
	// The receiver inside Describe() is Base, not User.
	fmt.Println("Describe sees Base only:", u.Describe())
}

File: go-backend-interview-prep/03_functions_and_methods/05_struct_embedding/main.py
# Python equivalent of Struct Embedding (compare with main.go)
# Go embedding = composition. Python uses inheritance.
# They look similar but have different semantics.


class Base:
    def __init__(self, id: int = 0):
        self.id = id

    def describe(self) -> str:
        return f"Base(id={self.id})"


class Logger:
    def __init__(self, prefix: str = ""):
        self.prefix = prefix

    def log(self, msg: str) -> None:
        print(f"[{self.prefix}] {msg}")


# --- Python inheritance (similar to Go embedding) ---
class User(Base):
    def __init__(self, id: int, name: str):
        super().__init__(id)
        self.name = name


# --- Multiple inheritance (similar to embedding multiple structs) ---
class Admin(User, Logger):
    def __init__(self, id: int, name: str, prefix: str, level: str):
        User.__init__(self, id, name)
        Logger.__init__(self, prefix)
        self.level = level


# --- Shadowing (override attribute) ---
class Employee(Base):
    def __init__(self, base_id: int, emp_id: str, name: str):
        super().__init__(base_id)
        self.emp_id = emp_id  # separate name to avoid confusion
        self.name = name


def main():
    # --- Basic inheritance ---
    print("--- Basic inheritance ---")
    u = User(id=1, name="Alice")
    print("u.id:", u.id)
    print("u.describe():", u.describe())  # inherited method
    print("u.name:", u.name)

    # --- Multiple inheritance ---
    print("\n--- Multiple inheritance ---")
    a = Admin(id=42, name="Bob", prefix="ADMIN", level="super")
    print("a.id:", a.id)
    print("a.name:", a.name)
    print("a.describe():", a.describe())
    a.log("system started")

    # --- Key difference ---
    print("\n--- Key difference ---")
    # In Python, describe() receives `self` (the Admin instance).
    # In Go, Describe() receives the embedded Base value.
    # Python has true inheritance; Go has composition.
    print("Python: describe() sees full object (inheritance)")
    print("Go: Describe() sees only Base (composition)")

    # --- Employee ---
    print("\n--- Employee ---")
    e = Employee(base_id=100, emp_id="EMP-001", name="Charlie")
    print("e.id:", e.id)          # from Base
    print("e.emp_id:", e.emp_id)  # separate field


if __name__ == "__main__":
    main()
